python/looker_sdk/rtl/auth_session.py

```diff
@@ -215,3 +219,110 @@ def _ok(self, response: transport.Response) -> str:
         if not response.ok:
             raise error.SDKError(response.value)
         return response.value.decode(encoding="utf-8")
+
+
+class CryptoHash:
+    def secure_random(self, byte_count: int) -> str:
+        return secrets.token_hex()
+
+    def sha256_hash(self, message: str) -> str:
+        value = hashlib.sha256()
+        value.update(bytes(message, "utf8"))
+        return value.hexdigest()
+
+
+class OAuthSession(AuthSession):
+    def __init__(
+        self,
+        *,
+        settings: api_settings.PApiSettings,
+        transport: transport.Transport,
+        deserialize: serialize.TDeserialize,
+        serialize: serialize.TSerialize,
+        crypto: CryptoHash,
+        version: str,
+    ):
+        super().__init__(settings, transport, deserialize, version)
+        self.crypto = crypto
+        self.serialize = serialize
+        config_data = self.settings.read_config()
+        for required in ["client_id", "redirect_uri", "looker_url"]:
+            if required not in config_data:
+                raise error.SDKError(f"Missing required configuration value {required}")
+
+        # would have prefered using setattr(self, required, ...) in loop above
+        # but mypy can't follow it
+        self.client_id = config_data["client_id"]
+        self.redirect_uri = config_data["redirect_uri"]
+        self.looker_url = config_data["looker_url"]
+        self.code_verifier = ""
+
+    def create_auth_code_request_url(self, scope: str, state: str) -> str:
+        self.code_verifier = self.crypto.secure_random(32)
+        code_challenge = self.crypto.sha256_hash(self.code_verifier)
+        params: Dict[str, str] = {
+            "response_type": "code",
+            "client_id": self.client_id,
+            "redirect_uri": self.redirect_uri,
+            "scope": scope,
+            "state": state,
+            "code_challenge_method": "S256",
+            "code_challenge": code_challenge,
+        }
+        path = urllib.parse.urljoin(self.looker_url, "auth")
+        query = urllib.parse.urlencode(params)
+        return f"{path}?{query}"
+
+    @attr.s(auto_attribs=True, kw_only=True)
+    class GrantTypeParams(model.Model):
+        client_id: str
+        redirect_uri: str
+
+    @attr.s(auto_attribs=True, kw_only=True)
+    class AuthCodeGrantTypeParams(GrantTypeParams):
+        code: str
+        code_verifier: str
+        grant_type: str = "authorization_code"
+
+    @attr.s(auto_attribs=True, kw_only=True)
+    class RefreshTokenGrantTypeParams(GrantTypeParams):
+        refresh_token: str
+        grant_type: str = "refresh_token"
```

Comment:
These types are undocumented in the API so there are no generated models for them. I've inlined them here because I only understand they're existence  in the python SDK as an implementation detail of this `OAuthSession` class. Alternatively they could go in `looker_sdk.rtl.model` as more "public" place but I assume they're undocumented in the API for a reason?

Reply:
We could possibly formalize "authorization credential sets" but I think it's still too early in the auth change cycle for that. For now I think checking the "is configured" function to find out if all required properties are provided on SDK initialization and providing details in the error message if not could be the way to do this. @dthorpe may have other ideas.